### DISCUSSION TITLE
Silencing a loader depth error

### DISCUSSION THREAD

**Discussion by michaelgmiller1** [sqlalchemy/sqlalchemy#10230] [Usage Questions] [answered]
We have a query that has a lot of nested loaders (12 levels deep). It's not elegant, and there most certainly is a better way to restructure the data for better performance. But in the mean time, we need to execute the query. We are hitting this warning:

```
            util.warn(
                "Loader depth for query is excessively deep; caching will "
                "be disabled for additional loaders.  Consider using the "
                "recursion_depth feature for deeply nested recursive eager "
                "loaders."
            )
```

Since the joins are not recursive, `recursion_depth` isn't applicable. And it doesn't seem possible to explicitly turn off `compiled_cache` to silence the warning. Is it possible to add a mechanism to forcibly disable caching, to avoid this warning?

**Comment by zzzeek** [comment]
Is there a reason you can't use the [python warning filter](https://docs.python.org/3/library/warnings.html#the-warnings-filter) here ?  That's part of the contract of warnings, that they are just trying to let you know something, and you certainly can control which warnings you care about.  When the eager loading turns off the caching here it's actually doing a lot more CPU work to generate these statements and related structures.

**Reply by michaelgmiller1** [reply]
Yes we absolutely could turn off the warning, but we'd prefer to do it on the context of a particular query. If we do have a deeply nested query we add in another case, we want to get the warning.

We want to declaratively state "we know this query will generate this warning, so don't warn for this particular query"

**Reply by zzzeek** [reply]
Oh ok on a per query basis.  Yes there's a way to disable the cache per query but probably not the warning.  So...would have to think of something.

**Reply by zzzeek** [reply]
it looks like there was an attempt to actually implement this, but it doesnt actually work because the execution options are not propagated to the additional loaders, assuming this `selectinload` you are using. looking

**Reply by zzzeek** [reply, marked as answer]
see #10231 which will propagate all user-facing execution options to secondary eager loaders when passed to `session.execute()`, `session.scalars()`.  so in 2.0.20 you would use:

    with Session(connection) as s:
        result = s.scalars(stmt, execution_options=dict(compiled_cache=None))

**Reply by CaselIT** [reply]
> We want to declaratively state "we know this query will generate this warning, so don't warn for this particular query"

You can use this context manager to contol the warning just on that query https://docs.python.org/3/library/warnings.html#warnings.catch_warnings

**Reply by zzzeek** [reply]
that manager is not thread safe, likely it's for test suites.  kind of a major caveat
